Enhance link control UI with rich URL previews

getdave opened the pull request:
![Screen Capture on 2021-06-16 at 15-29-55](https://user-images.githubusercontent.com/444434/122238049-bbcc5680-ceb7-11eb-9867-4c6a274914a1.gif)

## Description

**Please note**: currently the rich previews are only enabled within the post editor and only then within the format library implementation which is utilised by blocks such as `core/paragraph`. This intentional, in order to limit the surface area exposure to the change and allow it to "bed" in. We will enable across other areas of the editor in future subject to it working well where it is currently.

Building on https://github.com/WordPress/gutenberg/pull/18042 and https://github.com/WordPress/gutenberg/pull/31085, this PR wires things together to allow meta data about direct entry URLs (eg: `www.wordpress.org`) to be shown in the `<LinkControl>` UI (the UI that is used to add hyperlinks in Gutenberg).

This currently works in the Post Editor richtext only. It will not work (yet!) in 

* Site Editor
* Nav Editor
* Widget Editor

Closes https://github.com/WordPress/gutenberg/issues/31466

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Manual testing and component tests have been completed or added.

To manually test you should:

* Create some text in a paragraph block.
* Create several links (at least 3).
* Click between each link and see if the preview loads.
* Try different URLs and websites.
* Repeat but with your Network speed throttled. Especially try clicking rapidly between several previews to force bugs.

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/444434/121383960-c464e000-c93f-11eb-88e7-b0421d6131cf.mov


## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

getdave committed:
Add new helper to utilise new REST endpoint

getdave committed:
Provide docblock

getdave committed:
Get title tag contents from remote URL

getdave committed:
Create distinct data object for rich meta rather than rely on overiding suggestion result data.

getdave committed:
Add example of adding rich details to the link preview

getdave committed:
Add in description

getdave committed:
Remove url data to hook and prevent setting state on unmounted

getdave committed:
Avoid ref destructure

getdave committed:
Add loading animation

getdave committed:
Clear richdata on url change

getdave committed:
Avoid rich data in search items

getdave committed:
Account for preview layout variation requirements

getdave committed:
Only fetch if fetchRemoteUrlData is available

getdave committed:
Fix text overflow ellipsis

getdave committed:
Improve icon visual scaling

getdave commented:
cc @javierarce as you don't seem to be a member of this repo.

getdave committed:
Manage fetching as state

javierarce commented:
This works really nice! 

Some things I've noticed:

- I think we should cap the amount of text we show in the description. Two lines of texts (around 100 characters + ellipsis would work fine).
- I would also add some height constraints so that we don't show scrollbars and the popover has consistent height. If the image has a square format or doesn't fill up the width, we could show a light grey background around it.

<img width="402" alt="Screenshot 2021-05-26 at 13 57 54" src="https://user-images.githubusercontent.com/4933/119655250-aa98a880-be29-11eb-9930-5f28bb0d2ac4.png">

- I would also add a border-radius around the image (or, if we constraint the height, around the background cover).
- I would get rid of the external link icon. I understand why sometimes it is needed, but I think in this context is not necessary.

I also saw some a couple of bugs: 

#### The height of the popover should hug the content:

<img width="402" alt="Screenshot 2021-05-26 at 13 57 54" src="https://user-images.githubusercontent.com/4933/119656755-568ec380-be2b-11eb-8384-306b4ee8468b.png">

#### The icon is on top of the text (although this could be solved if we get rid of the icon)
<img width="381" alt="Screenshot 2021-05-26 at 14 01 46" src="https://user-images.githubusercontent.com/4933/119656746-542c6980-be2b-11eb-95f3-a8ccec59649a.png">

getdave committed:
Fix broken async/await code and restore test running

getdave committed:
Tidy code comments

getdave committed:
Apply correct spacing as per design provided

getdave committed:
Allow text to flow (almost) to edge of `Edit` button before ellipsis

getdave committed:
Constrain image height to 140px as per Figma design

getdave commented:
> This works really nice!

@javierarce Thanks! 👍 

I've made changes as per your feedback and updated the screenshot in the PR description. Let me know how this looks to you now.

### Some notes from me:

> * I think we should cap the amount of text we show in the description. Two lines of texts (around 100 characters + ellipsis would work fine).

Now capped to 2 lines using our new `<Text>` component. This seems to come from G2. Do you know if this is considered safe/stable to use? 

> * I would also add some height constraints so that we don't show scrollbars and the popover has consistent height. If the image has a square format or doesn't fill up the width, we could show a light grey background around it.

I've constrained the image to 140px which is as per the Figma. I will say I think it's too small as most images now cannot fill the space and we see a lot of gray background.

> * I would also add a border-radius around the image (or, if we constraint the height, around the background cover).

👍 Added to the background color.


> * I would get rid of the external link icon. I understand why sometimes it is needed, but I think in this context is not necessary.

Removed.


> I also saw some a couple of bugs:
> 
> #### The height of the popover should hug the content:
> <img alt="Screenshot 2021-05-26 at 13 57 54" width="402" src="https://user-images.githubusercontent.com/4933/119656755-568ec380-be2b-11eb-8384-306b4ee8468b.png">

I wasn't sure what you mean there.


> #### The icon is on top of the text (although this could be solved if we get rid of the icon)
> <img alt="Screenshot 2021-05-26 at 14 01 46" width="381" src="https://user-images.githubusercontent.com/4933/119656746-542c6980-be2b-11eb-95f3-a8ccec59649a.png">

The icon is currently 16x16. Why? Because the favicon we get back from most remote sites is 32x32 and low quality. By scaling down by 50% it gives the illusion of better quality. As it's smaller I've made it align with the top of the text as that seemed to look better. Do you agree?

@hellofromtonya perhaps in future we can look to parse an `og:` attribute or one of the other common "icon" meta values to grab a higher quality icon if it exists? I'd suggest if we do this then it's provided as another field so we can differentiate between them on the client side.

getdave committed:
Add icon, image and description

getdave committed:
Add icon, image and description

getdave committed:
Add tests to cover missing data edge cases

paaljoachim commented:
Hey Dave @getdave . I am not able to download the special Gutenberg plugin by going to the Checks tab. As I am not seeing the Gutenberg Build in the left sidebar area.

Btw I am very skeptical about seeing the featured image in the link control feature. As it does not matter what image the link shows. As one is mainly focused on getting in the link URL.

getdave commented:
The PR ~is~ was temporarily broken due to me making a mistake with a Git rebase. ~I'm working on restoring it~ I've fixed it.

> Btw I am very skeptical about seeing the featured image in the link control feature. As it does not matter what image the link shows. As one is mainly focused on getting in the link URL.

That's fair, although I disagree. I think when you consider it as part of a wider picture, it _does_ matter what image the link shows. 

The use case is that the image - along with title, description and icon - helps you to **_confirm/identify_** that you've entered the _correct_ URL. 

This is similar to what other editors (eg: Google Docs) do and I personally find it to be a nice affordance. That said, I'm open to a wide range of opinions 👍 

I've tried to address the rationale for this in [the Issue](https://github.com/WordPress/gutenberg/issues/31466).

getdave commented:
Just noticed a memory leak with component unmounting. Working on a fix.

javierarce commented:
Thanks for the changes, they are great.

What I meant by "hugging the content" was just having this:

<img width="396" alt="Screenshot 2021-05-27 at 14 58 58" src="https://user-images.githubusercontent.com/4933/119830210-527ea680-befc-11eb-8020-0ec7a2bd5ec5.png">

instead of this:

<img width="398" alt="Screenshot 2021-05-27 at 14 58 39" src="https://user-images.githubusercontent.com/4933/119830217-53afd380-befc-11eb-83f0-769f0b2f688d.png">

> The icon is currently 16x16. Why? Because the favicon we get back from most remote sites is 32x32 and low quality. By scaling down by 50% it gives the illusion of better quality. As it's smaller I've made it align with the top of the text as that seemed to look better. Do you agree?

Perfect, that's great.

javierarce commented:
> Btw I am very skeptical about seeing the featured image in the link control feature. As it does not matter what image the link shows. As one is mainly focused on getting in the link URL.

I agree with @getdave here, this improvement is about giving the users a little bit of extra context of the site they are linking. 

Also, HTML meta tags are one of the good things about the WWW, and this feature is an indirect way of supporting them.

getdave committed:
Fix cancel pending fetch when URL changes.

draganescu commented:
This is a very nice quality of life UX improvement. As a user I am always concerned with missing a letter and linking to some very similar honeypot domain. Also for long links (say a Medium.com article or an NYT article) seeing the featured image, along with the title and favicon, makes me instantly sure I did not link to a 404.

@paaljoachim I get what I think is your desire to simplify but the whole point imo of this is to quicken the process for users and visual indications that you can reference (presumably you've seen the article/website before linking) is top speed :) 

Great work @getdave !

paaljoachim commented:
Thank you everyone for very informative responses!

bph commented:
I am delighted that this is coming especially showing a featured image. As a visual person, it's so much easier to gauge if I grabbed the right link to share with the featured image. And I am not only thinking for the time when to publish an article. In a few months after publication, it will be very helpful, when I have to update an article. Hovering over the links with the image lets me see quickly if that still the link I thought it was. It's a fabulous gem of a feature that come to the block-editor. Thanks @getdave

getdave committed:
Remove isMounted anti pattern in favour of explictly cancelled promise pattern

See https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html

getdave committed:
Update comments to make reason for cancelling on URL change clearer

getdave committed:
Fix broken tests by disabling rich reviews in those tests which do not exercise this feature

getdave committed:
Move make cancelable util into own file

kevin940726 commented:
Can we rebase this PR now that #31763 is merged? It's a little bit hard to review now 😅

getdave commented:
Ill be able to rebase this PR on Wednesday.

getdave commented:
@kevin940726 This PR is now rebased.

getdave commented:
Could do with some advice on the e2e tests. They are largely failing because any test that adds a link to a direct URL now invokes a call to the new REST API endpoint which 404s and causes the test to fail. There are a lot of these tests.

Should I just methodically go through them all and add a mock response?

draganescu commented:
> Should I just methodically go through them all and add a mock response?

What other options are there? Is there a way to centrally hijack and mock them all?

kevin940726 commented on packages/block-editor/src/components/link-control/use-remote-url-data.js:
```diff
@@ -0,0 +1,88 @@
+		};
+	}, [] );
+
+	const cancelPendingFetch = useCallback( () => {
+		if ( cancelableFetch.current ) {
+			cancelableFetch.current.cancel();
+		}
+	}, [ cancelableFetch ] );
+
+	/**
+	 * Cancel any pending requests that were made for
+	 * stale URL values. If the component does not unmount
+	 * we must handle cancelling the current request if
+	 * the URL changes otherwise we will see stale data.
+	 */
+	useEffect( () => {
+		cancelPendingFetch();
+	}, [ url ] );
+
+	/**
+	 * Cancel any pending requests on "unmount"
+	 */
+	useEffect( () => {
+		return () => {
+			cancelPendingFetch();
+		};
+	}, [] );
```
I don't think any of these are necessary, we can just return a cleanup function in the effect below to call the cancel function instead.

getdave replied:
Weirdly the `<LinkPreview>` does not unmount when clicking between links. So if you click on a link and show the preview and then click immediately on another link, the popup doesn't unmount but rather simply renders down the tree.

This means that we need to actively cancel anything that's pending when the URL changes otherwise we will see stale data.

kevin940726 replied:
An easier and probably more straightforward way is to force remount the component whenever `url` changed via the `key` attribute. That way we can reset all the states and effects to prevent weird things from happening.

getdave replied:
> An easier and probably more straightforward way is to force remount the component whenever url changed via the key attribute

Done 👍

kevin940726 commented on packages/block-editor/src/components/link-control/use-remote-url-data.js:
```diff
@@ -0,0 +1,88 @@
+
+	useEffect( () => {
+		const fetchRichData = async () => {
+			setIsFetching( true );
+
+			// Clear the data if the URL changes to avoid stale data in hook consumer.
+			setRichData( null );
```
I believe this will cause two different renders since they are in a callback which React doesn't control. A better idea would be to leverage `useReducer` instead and encapsulate them together. It's also more extensible and we can add error handling too in the future.

kevin940726 commented on packages/block-editor/src/components/link-control/use-remote-url-data.js:
```diff
@@ -0,0 +1,88 @@
+			setRichData( null );
+
+			try {
+				cancelableFetch.current = makeCancelable(
```
I don't think we need to do this now that we can just use [`AbortController`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) to correctly _abort_ the request rather than _ignoring_ it. We might want to guard it from legacy browsers still though, but I don't think we have to polyfill it as we don't support IE anymore.

getdave replied:
Updated to use `AbortController`. Let me know what you think.

Main thing is we catch any attempts to update an unmounted component. 

Also how do we handle legacy browsers?

kevin940726 commented on packages/block-editor/src/components/link-control/link-preview.js:
```diff
@@ -7,49 +7,99 @@ import classnames from 'classnames';
+	const { richData, isFetching } = useRemoteUrlData( value?.url );
+
+	// Rich data may be an empty object so test for that.
+	const hasRichData = richData && Object.keys( richData ).length;
```
Should we just move this check inside the hook instead?

getdave replied:
I'm in two minds about this one.

If I've performed a fetch and it's returned an empty dataset wouldn't I want to know about that. If I return `null` from the hook if there's no data then I'm hiding that information from the consumer.

That could be a good thing or a bad thing.

How about another return value?

```js
const { richData, hasRichData, isFetching } = useRemoteUrlData( value?.url );
```

kevin940726 commented on packages/block-editor/src/components/link-control/link-preview.js:
```diff
@@ -7,49 +7,99 @@ import classnames from 'classnames';
 				'is-current': true,
+				'is-rich': hasRichData,
+				'is-fetching': !! isFetching,
+				'is-preview': true,
```
Is `has-preview` more suitable here?

getdave replied:
Actually not in this case as the "preview" is not related to the rich data preview but rather the "preview" of the link in general.

So the status is correct this _is_ a _preview_ rather than it _has_ a _preview_.

kevin940726 reviewed (changes requested):
Nice work! Some nitpicks and React specific suggestions :)

I noticed an annoying layout shift though when previewing the page. This would likely cause the user to mis-click, or even worse, click on the wrong things. It could also make the e2e tests unstable since puppeteer simulate the click event by splitting it into moving the cursor and pressing the button.

https://user-images.githubusercontent.com/7753001/120742716-04842700-c52a-11eb-9c6d-e28e906e259f.mp4

getdave commented:
> I noticed an annoying layout shift though when previewing the page.

🤔 @kevin940726 Hmmmm...what would you suggest here? We don't know if we will get the rich data or not and if we don't get data then we don't want to show the details area. It's tricky.

We could show a placeholder whilst it's loading but again these would need to disappear once things have loaded.

Any ideas welcome 😄

getdave committed:
Use AbortController to cancel requests rather than cancellable promise wrapper

getdave committed:
Force remount preview when the URL changes

Previously a URL might change and the component would not remount which could lead to edge cases.

Addresses https://github.com/WordPress/gutenberg/pull/31464#discussion_r645436478

getdave committed:
Implement simpler hook via abortable fetch

Based on https://codesandbox.io/s/dry-pond-43qy2?file=/src/link-control.js:187-201

getdave commented:
The main blocker here now is that multiple components consume `<LinkControl>`. In some of these showing a rich preview may not be desired (eg: Image block). There are also layout and styling issues with the `<LinkPreview>` component now that the markup has had to change to accommodate the rich preview.

An example of this is the Image block which consumes `<LinkControl>` but mostly deals with _internal_ URLs (which cannot have a rich preview) and really isn't set up with space to allow a full preview

https://user-images.githubusercontent.com/444434/120833658-f3530e80-c559-11eb-8431-a9c453cb09f0.mov

This is what's causing all the e2e tests to fail because it results in a request to the URL Details endpoint for an internal URL which ends up with a 404.

Needs much more thought about:

- How we stop requests going out for internal URLs without coupling the `<LinkControl>` to WordPress (it's currently a block editor component).
- How we _disable_ the rich feature by _default_.
- Enable only for certain scenarios.

I'm thinking the following route would be best:

- Add a prop to LinkControl called `richPreviews` and default to `false`.
- Pass said prop down to `LinkPreview` and short-circuit the fetching by doing something like this
```js
// Avoid fetching if rich previews are not desired.
const maybeRemoteURL = richPreviews ? value?.url : null;

const { richData, isFetching } = useRemoteUrlData( maybeRemoteURL );
```
- Set the `richPreviews` prop of `<LinkControl>` to true only in the places where we are consuming the component and we know we want to show rich previews (eg: `RichText` component).

getdave committed:
Fix to ensure correct handling of aborted vs standard fetch errors

Avoid situation where requests which 404 do not result in the fetching UI being reset.

getdave committed:
Add test to cover resetting fetching state if rich data requests fails

kevin940726 commented on packages/block-editor/src/components/link-control/use-remote-url-data.js:
```diff
@@ -0,0 +1,86 @@
+		};
+	}, [] );
+
+	function abortableFetch( theUrl, signal ) {
```
Don't need to create this function since `apiFetch` (`fetchRemoteUrlData`) should be able to accept the `signal` property from `AbortController`.

```js
fetchRemoteUrlData(url, { signal })
```

kevin940726 commented on packages/block-editor/src/components/link-control/use-remote-url-data.js:
```diff
@@ -0,0 +1,86 @@
+		// and the fetching util is available. In some editors
+		// there may not be such a util.
+		if ( url?.length && fetchRemoteUrlData ) {
+			fetchRichData();
```
We can just call the function and return the effect here, so that we don't have to declare controller using a let, we can just use `const`.

kevin940726 reviewed (commented):
The whole `useRemoteUrlData` should be only containing:

1. a `useReducer` to store and set the states
2. a `useSelect` to get the `fetchRemoteUrlData` function
3. a `useEffect` to call the api, including the `AbortController` and effect clean up function

kevin940726 commented:
`richPreviews` prop sounds good (though I would probably rename it to `hasRichPreviews` indicating that it's a boolean value)!

getdave committed:
Simplify hook implementation

getdave committed:
Guard for legacy browsers

getdave committed:
Refactor to useReducer

getdave commented:
@javierarce 👋 I have a question for you 😄. @kevin940726 has spotted a rendering issue whereby the popup will "jump" once the rich data has loaded. See below:

https://user-images.githubusercontent.com/444434/121005108-5aa1d600-c787-11eb-9f1b-86b56ddcffd8.mov

There's no easy solution to this as the positioning of the popover is handled automatically based on several variables such as window size, content size...etc.

This also relates to [this issue whereby the Block toolbar is obscured by the popover](https://github.com/WordPress/gutenberg/pull/31464#issuecomment-856120116).

My suggestion would be that if rich previews are enabled then we render some placeholders (eg: gray boxes) for the image and description. When the preview is "loading" then we have them gently pulse. Once the rich data request comes back (or not) we then have 2 scenarios:

1. We have rich data - then render the image and description and remove the placeholders 
2. We don't have rich data - then leave the placeholders in place but stop them pulsing.

I like `#2` above because it both stops the layout shift and it provides a good indication that the reason you haven't got a rich preview is that data was _missing_ rather than because something went wrong.

It would be great to know if you're happy with this idea.

## Update

Here's a loading state I created

![Screen Capture on 2021-06-08 at 09-48-47 (1)](https://user-images.githubusercontent.com/444434/121154327-cb57f980-c83e-11eb-84e9-8220893111af.gif)

![Screen Capture on 2021-06-08 at 10-46-52](https://user-images.githubusercontent.com/444434/121163732-f0506a80-c846-11eb-8fc0-f5d6f322ab5f.gif)

getdave committed:
Disable rich previews by default and enable only in inline text by default

kevin940726 commented on packages/block-editor/src/components/link-control/test/index.js:
```diff
@@ -1713,3 +1733,312 @@ describe( 'Post types', () => {
+			render( <LinkControl value={ selectedLink } />, container );
+		} );
+
+		// mockFetchRemoteUrlData resolves on next "tick" of event loop
+		await act( async () => {
+			await eventLoopTick();
+		} );
```
IMO, _when_ the link previews show is an implementation detail, we only care about whether they show or not. Instead, we can use the `findBy*` queries from `@testing-library/react` to query the resolved UI.

getdave replied:
I had no idea we had access to RTL. I would _much_ prefer to use those utils. That said, the pattern in these tests currently is to use the `eventLoopTick()`. Could we do this in a followup to avoid making this PR any larger?

kevin940726 replied:
Oh I didn't know that, sure! Absolutely.

kevin940726 commented on packages/editor/src/components/provider/use-block-editor-settings.js:
```diff
@@ -102,8 +102,8 @@ function useBlockEditorSettings( settings, hasTemplate ) {
 				fetchLinkSuggestions( search, searchOptions, settings ),
-			__experimentalFetchRemoteUrlData: ( url ) =>
-				fetchRemoteUrlData( url ),
+			__experimentalFetchRemoteUrlData: ( url, options ) =>
+				fetchRemoteUrlData( url, options ),
```
Nitpicking: we can probably just do this?

```suggestion
			__experimentalFetchRemoteUrlData: fetchRemoteUrlData,
```

kevin940726 commented on packages/block-editor/src/components/link-control/index.js:
```diff
@@ -115,6 +115,7 @@ function LinkControl( {
 	suggestionsQuery = {},
 	noURLSuggestion = false,
 	createSuggestionButtonText,
+	richPreviews = false,
```
I still like `hasRichPreviews` better 😅, any reason why you went for just `richPreviews`?

getdave replied:
To me the word "has" implies something is "currently" X or Y. When you write a component or pass a prop you are thinking about what it _will_ be in the future once it has rendered.

Therefore I was thinking for simplicity just saying:

> `richPreviews` are they "on" or "off"?

This is not super important to me however, so I'm happy to change if you feel it's better as "has".

getdave replied:
I changed it 👍

kevin940726 replied:
I always use `is`, `has`, `should`, or `will` prefix to indicate the variable is a boolean. That's just my personal preference though!

> When you write a component or pass a prop you are thinking about what it will be in the future once it has rendered.

That's an interesting interpretation! I don't think it that way though. Normally I would think of a component as a function for UI, the arguments are props, and the return value is the UI state of those props. If we put `state` aside, then it's just a pure function which maps props to a UI, so I won't say that it's "will", but "has".

But again, it's just a preference thing, I respect your decision 👍.

getdave committed:
Update packages/editor/src/components/provider/use-block-editor-settings.js

Co-authored-by: Kai Hao <kevin830726@gmail.com>

getdave committed:
Update to hasRichPreviews

Address feedback from https://github.com/WordPress/gutenberg/pull/31464#discussion_r646582909

getdave committed:
Use correct prop to enable rich previews in rich text

getdave commented:
Failing e2e test can be simulated locally by running

```
THROTTLE_CPU=4 npm run test-e2e:watch -- --puppeteer-interactive packages/e2e-tests/specs/editor/various/links.test.js
```

I believe the issue is that with certain screen dimensions we're covering the "Unlink" button with the popover preview.

![Screen Capture on 2021-06-08 at 09-20-10](https://user-images.githubusercontent.com/444434/121149718-ce50eb00-c83a-11eb-823c-538c595f3011.gif)

I wonder if we can discuss a solution to this @javierarce?

getdave committed:
Fix e2e test to allow unlinking when rich data obscures block toolbar

javierarce commented:
> We don't have rich data - then leave the placeholders in place but stop them pulsing.
> I like #2 above because it both stops the layout shift and it provides a good indication that the reason you haven't got a rich preview is that data was missing rather than because something went wrong.

Would it be possible to load the placeholders by default, disable the positioning system for the popover, and then hide the placeholders from the content that couldn't get loaded?

If that's not possible, I guess your solution is the best thing we can do.

Regarding the issue with the "unlink" button:

> I believe the issue is that with certain screen dimensions we're covering the "Unlink" button with the popover preview.

I understand that moving the popover so that the toolbar is always visible is a tough task. Also, since the popover is fixed and relative to the linked text, I think one solution could be adding an unlink option at the bottom of the popover:

![image](https://user-images.githubusercontent.com/4933/121180537-40d4c180-c861-11eb-975e-7359470513e9.png)

getdave commented:
> Would it be possible to load the placeholders by default

It would, but...

> disable the positioning system for the popover,

...I don't think we can disable this. It's designed to keep the popover within the viewport. It might need to be "toolbar aware" but that's way outside the scope of this PR 😟 

> and then hide the placeholders from the content that couldn't get loaded?

The only concern I have with this is that we'll still have the "jump" when the placeholders get removed. 

> I think one solution could be adding an unlink option at the bottom of the popover:

Yeh I agree this is a good idea. Do you think we ought to include the icon used in the toolbar or is that too much?

<img width="425" alt="Screen Shot 2021-06-08 at 14 10 16" src="https://user-images.githubusercontent.com/444434/121190815-521ecd80-c863-11eb-8ba8-bf36d8a0f351.png">

javierarce commented:
> ..I don't think we can disable this. It's designed to keep the popover within the viewport. It might need to be "toolbar aware" but that's way outside the scope of this PR 😟

In that case, I say let's go with the solution you propose.

> Yeh I agree this is a good idea. Do you think we ought to include the icon used in the toolbar or is that too much?

I tried that, but I think showing the icon would bring too much attention to that link. Also, since we don't use the red color for other destructive actions like "Remove block", I would say to style it in black (with a blue hover, like the toolbar options).

getdave committed:
Add permenant missing data placeholders with loading animation

getdave commented:
@javierarce I've added the `Unlink` button and put the placeholders in place.

https://user-images.githubusercontent.com/444434/121216117-8355c880-c878-11eb-8e0e-5fa3f9ad6026.mp4



@kevin940726 You'll notice that adding `Unlink` was not a simple process and involved updating the `@wordpress/format-library` package to handle this for rich text.

Note, however, that this `Unlink` button will now appear for _all_ instances of `<LinkControl>` and it's quite possible that consumers of that component will not expect a `null` value to be passed as the `value`. This could cause a lot of breakages.

Options

1. Make the `unlink` button optional via a prop or something.
2. Accept the breakages and try and fix the ones we control.

If you can think of a better way let me know.

Moreover, I"m thinking that we should break out the addition of `Unlink` into a separate PR as it's really unrelated to this one and it's starting to get rather large. What do you think?

getdave committed:
Provide docblock

getdave committed:
Add tests for rich data and resolve exposed bugs

getdave committed:
Updates to use visually truncated description text

getdave committed:
Clip image height, center and provide border-radius and background

getdave committed:
Allow for passing options to fetchRemoteUrlData

This enables the usage of AbortController signal

getdave committed:
Fix tests

getdave commented:
Ok, I think `Unlink` adds too much complexity to this PR. I'm going to:

- [x] Pull the addition of `Unlink` button into it's own PR. [**Done**](https://github.com/WordPress/gutenberg/pull/32541).
- [x] Revert the changes in this PR. **Done**.
- [ ] Merge this PR.


Once merged we can followup with the addition of `Unlink` as that raises a lot of additional questions about how best to extend the component to allow for it.

getdave committed:
Update snapshots

javierarce commented:
I think this looks and works nice, but I think we should address the hiding of the placeholders at some point, that would make this look really great.

kevin940726 commented on packages/block-editor/src/components/link-control/make-cancelable.js:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @param {Promise} promise the Promise to make cancelable
+ */
+const makeCancelable = ( promise ) => {
```
Do we still need this?

getdave replied:
Yeh it's used in the create item flow (not introduced by this PR). But we can probably revert a couple of changes that move it to a separate file?

getdave replied:
Done

kevin940726 commented on packages/core-data/src/fetch/__experimental-fetch-remote-url-data.js:
```diff
@@ -15,8 +15,8 @@ import { addQueryArgs, prependHTTP } from '@wordpress/url';
  * @async
- * @param {string} url
- *
+ * @param {string} url the URL to request details from.
+ * @param {Object?} options any options to pass to the underlying fetch.
```
```suggestion
 * @param {string}  url     the URL to request details from.
 * @param {Object?} options any options to pass to the underlying fetch.
```

kevin940726 reviewed (approved):
Looking good! Let's address other issues in follow-up PRs 👍

getdave committed:
Update doc block

Co-authored-by: Kai Hao <kevin830726@gmail.com>

getdave committed:
Restore makecancellable to a local util

getdave commented:
> I think this looks and works nice, but I think we should address the hiding of the placeholders at some point, that would make this look really great.

@javierarce and I discussed this and we'll continue all further UI enhancements in follow-up Issues.

youknowriad commented:
For the future, it would be cool if we try to limit the use of snapshots in tests (specially big ones). These tests are harder to debug and change to often when implementation details do.

kevin940726 commented:
> For the future, it would be cool if we try to limit the use of snapshots in tests (specially big ones). These tests are harder to debug and change to often when implementation details do.

@youknowriad I strongly agree with this. Snapshots have their advantages, but often times we were overusing them.

A rule of thumb of mine is that if the snapshot doesn't look good when inlining them in the test (`toMatchInlineSnapshot()`), then it's probably too long to include them in a separate file anyway. But then again, why don't we just use `toMatchInlineSnapshot()` for that? `toMatchSnapshot()` has very few use cases to me IMHO.

I believe we haven't really reached a consensus about this though, it's still being mentioned/recommended in our [doc for testing](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/testing-overview.md#use-cases). It would be good if we can start a new issue, reach a consensus about snapshot testing in general, and update the doc.

draganescu commented:
Extraordinary work here @getdave ! For me this is a feature which although contained by a small box and only visible on links, it makes the whole experience immensely up to date: rich previews is the expected interaction in a software landscape focused more and more on refined interactions that confirm expected outcomes. 🎉 👏🏻
